tests/52user-directory.pl

```diff
@@ -0,0 +1,449 @@
+         )
+      })->then( sub {
+         get_user_dir_synced( $user, $displayname )
+      })->then( sub {
+         my ( $body ) = @_;
+
+         any { $_->{user_id} eq $user->user_id } @{ $body->{results} }
+            and die "user in list";
+
+         pass "User not in directory after history_visibility set to shared";
+
+         Future->done( 1 )
+      });
+   };
+
+test "User in remote room don't appear in user directory after server left room",
+   requires => [ local_user_fixture(), remote_user_fixture() ],
+
+   check => sub {
+      my ( $creator, $remote ) = @_;
+
+      my ( $room_id, $displayname );
+
+      create_and_set_random_displayname( $creator )
+      ->then( sub {
+         ( $displayname ) = @_;
+
+         matrix_create_room( $creator,
+            preset => "public_chat", invite => [ $remote->user_id ],
+         )
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         log_if_fail "room_id", $room_id;
+
+         matrix_join_room( $remote, $room_id );
+      })->then( sub {
+         get_user_dir_synced( $remote, $displayname )
+      })->then( sub {
+         my ( $body ) = @_;
+
+         log_if_fail "Body", $body;
+
+         any { $_->{user_id} eq $creator->user_id } @{ $body->{results} }
+            or die "user not in list";
+
+         matrix_leave_room( $remote, $room_id );
+      })->then( sub {
+         get_user_dir_synced( $remote, $displayname )
+      })->then( sub {
+         my ( $body ) = @_;
+
+         log_if_fail "Body", $body;
+
+         any { $_->{user_id} eq $creator->user_id } @{ $body->{results} }
+            and die "user in list";
+
+         Future->done( 1 );
+      });
+   };
+
+test "User directory correctly update on display name change",
+   requires => [ local_user_fixture() ],
+
+   check => sub {
+      my ( $user ) = @_;
+
+      my ( $room_id, $displayname );
+
+      create_and_set_random_displayname( $user )
+      ->then( sub {
+         ( $displayname ) = @_;
+
+         log_if_fail "First displayname", $displayname;
+
+         matrix_create_room( $user,
+            preset => "public_chat",
+         )
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         get_user_dir_synced( $user, $displayname )
+      })->then( sub {
+         my ( $body ) = @_;
+
+         log_if_fail "Body", $body;
+
+         any { $_->{user_id} eq $user->user_id } @{ $body->{results} }
+            or die "user not in list";
+
+         create_and_set_random_displayname( $user )
+      })->then( sub {
+         ( $displayname ) = @_;
+
+         log_if_fail "Second displayname", $displayname;
+
+         get_user_dir_synced( $user, $displayname )
+      })->then( sub {
+         my ( $body ) = @_;
+
+         log_if_fail "Second Body", $body;
+
+         any { $_->{user_id} eq $user->user_id } @{ $body->{results} }
+            or die "user not in list";
+
+         Future->done( 1 )
+      });
+   };
+
+sub create_and_set_random_displayname
```

Comment:
Despite its convenience, this all-in-one function does feel a bit awkward to use, for having to capture the `$displayname` each time. I wonder about splitting it in two, so you can

    my $displayname = generate_random_displayname();
    ...
    matrix_set_displayname( $user, $displayname )->then ...